File: ts-host/src/native/host-tree.ts
export type TreeEntry = { name: string; kind: 'branch' | 'leaf' };
export interface TreeProvider<T = unknown> {
  list(path: readonly string[]): Iterable<TreeEntry>;
  read(path: readonly string[]): T;
}

type Cache = { provider: TreeProvider; branches: Map<string, TreeEntry[]>; leaves: Map<string, unknown> };
const key = (path: readonly string[]) => path.join('/');

/** Stable read-only view: each observed branch or leaf is fetched at most once. */
export class LazyDict<T = unknown> {
  readonly __natlangLazyDict = true;
  private readonly cache: Cache;
  constructor(provider: TreeProvider<T>, readonly label = 'host tree', readonly path: readonly string[] = [], cache?: Cache) {
    this.cache = cache ?? { provider, branches: new Map(), leaves: new Map() };
  }
  entries(): TreeEntry[] {
    const at = key(this.path), found = this.cache.branches.get(at);
    if (found) return found;
    const names = new Set<string>(), entries: TreeEntry[] = [];
    for (const entry of this.cache.provider.list(this.path)) {
      if (!entry.name || /[/\\]/.test(entry.name) || ['.', '..'].includes(entry.name) || entry.name.startsWith('$') ||
          !['branch', 'leaf'].includes(entry.kind)) throw new TypeError(`invalid host-tree entry ${JSON.stringify(entry)}`);
      if (names.has(entry.name)) throw new TypeError(`duplicate host-tree entry ${entry.name}`);
      names.add(entry.name); entries.push({ name: entry.name, kind: entry.kind });
    }
    entries.sort((a, b) => a.name < b.name ? -1 : a.name > b.name ? 1 : 0);
    this.cache.branches.set(at, entries); return entries;
  }
  child(name: string): LazyDict<T> | T {
    const entry = this.entries().find(item => item.name === name);
    if (!entry) throw new RangeError(`no such host-tree entry: ${name}`);
    const path = [...this.path, name], at = key(path);
    if (entry.kind === 'branch') return new LazyDict(this.cache.provider as TreeProvider<T>, this.label, path, this.cache);
    if (!this.cache.leaves.has(at)) this.cache.leaves.set(at, this.cache.provider.read(path));
    return this.cache.leaves.get(at) as T;
  }
}

export function isLazyDict(value: unknown): value is LazyDict {
  return !!value && typeof value === 'object' && (value as { __natlangLazyDict?: unknown }).__natlangLazyDict === true;
}

export class MemoryTreeProvider<T = unknown> implements TreeProvider<T> {
  private readonly leaves = new Map<string, T>();
  constructor(leaves: Record<string, T>) {
    for (const [raw, value] of Object.entries(leaves)) {
      const parts = raw.replaceAll('\\', '/').split('/').filter(Boolean);
      if (!parts.length || raw.startsWith('/') || parts.some(part => part === '.' || part === '..'))
        throw new TypeError(`invalid host-tree path: ${raw}`);
      this.leaves.set(parts.join('/'), value);
    }
  }
  list(path: readonly string[]): TreeEntry[] {
    const prefix = path.length ? `${key(path)}/` : '', children = new Map<string, 'branch' | 'leaf'>();
    for (const leaf of this.leaves.keys()) {
      if (!leaf.startsWith(prefix)) continue;
      const rest = leaf.slice(prefix.length), [name] = rest.split('/');
      if (!name) continue;
      const kind = rest.includes('/') ? 'branch' : 'leaf';
      if (children.get(name) !== 'branch') children.set(name, kind);
    }
    return [...children].map(([name, kind]) => ({ name, kind }));
  }
  read(path: readonly string[]): T {
    const at = key(path);
    if (!this.leaves.has(at)) throw new RangeError(`no such host-tree leaf: ${at}`);
    return this.leaves.get(at)!;
  }
}

export const lazyDict = <T>(leaves: Record<string, T>, label = 'lazy dictionary') =>
  new LazyDict(new MemoryTreeProvider(leaves), label);

File: natlang/host_tree.py
"""General read-only, lazily observed trees supplied by an application host."""
from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Iterable, Mapping, Protocol


@dataclass(frozen=True)
class TreeEntry:
    name: str
    kind: str  # branch | leaf


class TreeProvider(Protocol):
    def list(self, path: tuple[str, ...]) -> Iterable[TreeEntry]: ...
    def read(self, path: tuple[str, ...]) -> Any: ...


class _TreeCache:
    def __init__(self, provider: TreeProvider):
        self.provider = provider
        self.branches: dict[tuple[str, ...], tuple[TreeEntry, ...]] = {}
        self.leaves: dict[tuple[str, ...], Any] = {}


class LazyDict:
    """Host representation of Dict<T>; each observed branch or leaf is fetched once."""

    def __init__(self, provider: TreeProvider, *, label: str = "host tree",
                 _cache: _TreeCache | None = None, _path: tuple[str, ...] = ()):
        self._cache = _cache or _TreeCache(provider)
        self._path = _path
        self.label = label

    @property
    def path(self) -> tuple[str, ...]:
        return self._path

    def entries(self) -> tuple[TreeEntry, ...]:
        if self._path not in self._cache.branches:
            entries, names = [], set()
            for raw in self._cache.provider.list(self._path):
                entry = raw if isinstance(raw, TreeEntry) else TreeEntry(str(raw["name"]), str(raw["kind"]))
                if (not entry.name or "/" in entry.name or "\\" in entry.name or entry.name in (".", "..") or
                        entry.name.startswith("$") or entry.kind not in ("branch", "leaf")):
                    raise ValueError(f"invalid host-tree entry: {entry!r}")
                if entry.name in names:
                    raise ValueError(f"duplicate host-tree entry: {entry.name}")
                names.add(entry.name); entries.append(entry)
            self._cache.branches[self._path] = tuple(sorted(entries, key=lambda item: item.name))
        return self._cache.branches[self._path]

    def child(self, name: str):
        entry = next((item for item in self.entries() if item.name == name), None)
        if entry is None:
            raise KeyError(name)
        path = self._path + (name,)
        if entry.kind == "branch":
            return LazyDict(self._cache.provider, label=self.label, _cache=self._cache, _path=path)
        if path not in self._cache.leaves:
            self._cache.leaves[path] = self._cache.provider.read(path)
        return self._cache.leaves[path]

    def __deepcopy__(self, memo):
        return self

    def __repr__(self):
        at = "/".join(self._path)
        suffix = f", {at!r}" if at else ""
        return f"LazyDict({self.label!r}{suffix})"


class MemoryTreeProvider:
    """Portable provider over slash-separated leaf paths."""

    def __init__(self, leaves: Mapping[str, Any]):
        self.leaves = {}
        for raw, value in leaves.items():
            path = tuple(part for part in str(raw).replace("\\", "/").split("/") if part)
            if not path or str(raw).startswith("/") or any(part in (".", "..") for part in path):
                raise ValueError(f"invalid host-tree path: {raw}")
            self.leaves[path] = value
        for path in self.leaves:
            if any(path[:end] in self.leaves for end in range(1, len(path))):
                raise ValueError(f"host-tree leaf is also a branch: {'/'.join(path)}")

    def list(self, path: tuple[str, ...]):
        children = {}
        for leaf in self.leaves:
            if leaf[:len(path)] != path or len(leaf) == len(path):
                continue
            name = leaf[len(path)]
            kind = "leaf" if len(leaf) == len(path) + 1 else "branch"
            if children.get(name) == "branch":
                continue
            children[name] = kind
        return [TreeEntry(name, kind) for name, kind in children.items()]

    def read(self, path: tuple[str, ...]):
        if path not in self.leaves:
            raise KeyError("/".join(path))
        return self.leaves[path]


def lazy_dict(leaves: Mapping[str, Any], *, label: str = "lazy dictionary") -> LazyDict:
    return LazyDict(MemoryTreeProvider(leaves), label=label)
